components/ws-daemon/nsinsider/main.go

```diff
@@ -254,6 +256,101 @@ func main() {
 					return nil
 				},
 			},
+			{
+				Name:  "setup-pair-veths",
+				Usage: "set up a pair of veths",
+				Flags: []cli.Flag{
+					&cli.IntFlag{
+						Name:     "target-pid",
+						Required: true,
+					},
+				},
+				Action: func(c *cli.Context) error {
+					containerIf, vethIf, cethIf := "eth0", "veth0", "ceth0"
+					mask := net.IPv4Mask(255, 255, 255, 0)
+					vethIp := net.IPNet{
+						IP:   net.IPv4(10, 0, 5, 1),
+						Mask: mask,
+					}
+					masqueradeAddr := net.IPNet{
+						IP:   vethIp.IP.Mask(mask),
+						Mask: mask,
+					}
+					iptablesCmd := "/usr/sbin/iptables"
+					targetPid := c.Int("target-pid")
+
+					veth := &netlink.Veth{
+						LinkAttrs: netlink.LinkAttrs{
+							Name:  vethIf,
+							Flags: net.FlagUp,
+							MTU:   1500,
+						},
+						PeerName:      cethIf,
+						PeerNamespace: netlink.NsPid(targetPid),
+					}
+					if err := netlink.LinkAdd(veth); err != nil {
+						return xerrors.Errorf("link %q-%q netns failed: %v", vethIf, cethIf, err)
+					}
+
+					vethLink, err := netlink.LinkByName(vethIf)
+					if err != nil {
+						return xerrors.Errorf("cannot found %q netns failed: %v", vethIf, err)
+					}
+					if err := netlink.AddrAdd(vethLink, &netlink.Addr{IPNet: &vethIp}); err != nil {
+						return xerrors.Errorf("failed to add IP address to %q: %v", vethIf, err)
+					}
+					if err := netlink.LinkSetUp(vethLink); err != nil {
+						return xerrors.Errorf("failed to enable %q: %v", vethIf, err)
+					}
+
+					cmd := exec.Command(iptablesCmd, "-t", "nat", "-A", "POSTROUTING", "-s", masqueradeAddr.String(), "-o", containerIf, "-j", "MASQUERADE")
```

Comment:
`nsinsider` should never exec out like this because it runs as actual root (i.e. has all capabilities on the node), yet executes within the mount namespace of the container. Hence, we're running a user-supplied `iptables` as actual root. It's easy to see how this is a security issue.

If we want to keep using iptables, we might have to resort to the iptables command (unfortunately netfilter doesn't have too great a userspace API). To do this we must:
- ensure within nsinsider that we're running in the original ws-daemon mount namespace. This is a safety precaution so that we don't inadvertently execute `iptables` from a different context in the future. I'm not sure how to do this reliably though :/ 
- use `iptables` that we ship within `ws-daemon`.


Alternatively, we could use `nftables` (the successor to iptables) for which [better library support exists](https://github.com/google/nftables).

Reply:
@csweichel  Thanks for the great findings. You are absolutely right. So `iptablesCmd` just happens to exist depending on the user, right? And thanks for the great suggestions. I'm giving it some thought, I didn't know about `nftables` so I'll give it a bit of a try.
I looked into `iptables` command too, but it looks like all I can do is use the command.